robotsim.py

```diff
@@ -48,118 +48,198 @@
+                generate_map()
+                self.set_position(x2,y2,angle)
 
     def move_backward(self):
-        angle = self.w
-        x1 = self.x 
-        y1 = self.y 
-        rad = math.radians(angle)
-        x2 = x1 - round(math.cos(rad))
-        y2 = round(math.sin(rad)) + y1
-        generate_map()
-        # Check collisions
-        px1 = self.x + int(pixel_constant * 0.5) - round(math.cos(rad)*self.size*0.5) 
-        py1 = self.y + int(pixel_constant * 0.5) + round(math.sin(rad)*self.size*0.5)
-        for i in range(-int(self.size*0.5)+1,int(self.size*0.5)-1): 
-            px2 = round(math.cos(math.radians(90+angle)) * i) + px1
-            py2 = py1 - round(math.sin(math.radians(90+angle)) * i)
-            is_wall = gameDisplay.get_at((px2,py2)) == colors['black']
-            if is_wall:
-                return
-        self.set_position(x2,y2,angle)
+        #TODO
+        pass 
     
     def rotate_right(self):
-        generate_map()
-        self.set_position(self.x,self.y,self.w - 1)
+        if 1 <= self.dir <= 3:
+            self.dir -= 1
+        else:
+            self.dir = 3
+        for _ in range(30):
+            generate_map()
+            self.set_position(self.x,self.y,self.w - 3)
 
     def rotate_left(self):
-        generate_map()
-        self.set_position(self.x,self.y,self.w + 1)
-
-    def get_distance(self,angle):
-        rad = math.radians(angle)
-        x1 = self.x + int(pixel_constant * 0.5) + round(math.cos(rad)*self.size*0.5)
-        y1 = self.y + int(pixel_constant * 0.5) - round(math.sin(rad)*self.size*0.5)
-        for i in range(self.sensor_range):
-            x2 = int(math.cos(math.radians(angle)) * i) + x1
-            y2 = y1 - int(math.sin(math.radians(angle)) * i)
-            x2 = 0 if x2 < 0 else display_width if x2 > display_width else x2
-            y2 = 0 if y2 < 0 else display_height if y2 > display_height else y2
-            is_wall = gameDisplay.get_at((x2,y2)) == colors['black']
-            pygame.draw.line(gameDisplay, red, (x1, y1), (x2, y2))
-            if is_wall:
-                pygame.display.update()
-                clock.tick(120)
-                return i
-        return self.sensor_range
-
-    def ultrasonic_forward(self):
-        return self.get_distance(self.w)
-
-    def ultrasonic_left(self):
-        return self.get_distance(self.w + 90)
-
-    def ultrasonic_back(self):
-        return self.get_distance(self.w + 180)
-
-    def ultrasonic_right(self):
-        return self.get_distance(self.w + 270)
-
-    def get_color(self):
-        x = self.x + int(pixel_constant * 0.5)
-        y = self.y + int(pixel_constant * 0.5)
-        generate_map()
-        c = gameDisplay.get_at((x,y))
-        self.set_position(self.x,self.y,self.w)
-        for name, value in colors.items():
-            if value == c:
-                return name
+        if 0 <= self.dir <= 2:
+            self.dir += 1
+        else:
+            self.dir = 0
```

Comment:
You could do something like this : (self.dir + 1) % 4.

Reply:
Just to simplify it and avoid using the if/else conditions

Reply:
Done! Great idea.